cmd/dep/gb_importer.go

```diff
@@ -0,0 +1,169 @@
+	Path string `json:"path,omitempty"`
+}
+
+func (i *gbImporter) Name() string {
+	return "gb"
+}
+
+func (i *gbImporter) HasDepMetadata(dir string) bool {
+	// gb stores the manifest in the vendor tree
+	var m = filepath.Join(dir, "vendor", "manifest")
+	if _, err := os.Stat(m); err != nil {
+		return false
+	}
+
+	return true
+}
+
+func (i *gbImporter) Import(dir string, pr gps.ProjectRoot) (*dep.Manifest, *dep.Lock, error) {
+	err := i.load(dir)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return i.convert(pr)
+}
+
+// load the gb manifest
+func (i *gbImporter) load(projectDir string) error {
+	i.logger.Println("Detected gb manifest file...")
+	var err error
+	var mf = filepath.Join(projectDir, "vendor", "manifest")
+	if i.verbose {
+		i.logger.Printf("  Loading %s", mf)
+	}
+
+	var buf []byte
+	if buf, err = ioutil.ReadFile(mf); err != nil {
+		return errors.Wrapf(err, "Unable to read %s", mf)
+	}
+	if err = json.Unmarshal(buf, &i.manifest); err != nil {
+		return errors.Wrapf(err, "Unable to parse %s", mf)
+	}
+
+	return nil
+}
+
+// convert the gb manifest into dep configuration files.
+func (i *gbImporter) convert(pr gps.ProjectRoot) (*dep.Manifest, *dep.Lock, error) {
+	i.logger.Println("Converting from gb manifest...")
+
+	manifest := &dep.Manifest{
+		Constraints: make(gps.ProjectConstraints),
+	}
+
+	lock := &dep.Lock{}
+
+	for _, pkg := range i.manifest.Dependencies {
+		pc, lp, err := i.convertOne(pkg)
+		if err != nil {
+			return nil, nil, err
+		}
+		manifest.Constraints[pc.Ident.ProjectRoot] = gps.ProjectProperties{Source: pc.Ident.Source, Constraint: pc.Constraint}
+		lock.P = append(lock.P, lp)
+	}
+
+	return manifest, lock, nil
+}
+
+func (i *gbImporter) convertOne(pkg gbDependency) (pc gps.ProjectConstraint, lp gps.LockedProject, err error) {
+	if pkg.Importpath == "" {
+		err = errors.New("Invalid gb configuration, package import path is required")
+		return
+	}
+
+	/*
+		gb's vendor plugin (gb vendor), which manages the vendor tree and manifest
+		file, supports fetching by a specific tag or revision, but if you specify
+		either of those it's a detached checkout and the "branch" field is HEAD.
+		The only time the "branch" field is not "HEAD" is if you do not specify a
+		tag or revision, otherwise it's either "master" or the value of the -branch
+		flag
+
+		This means that, generally, the only possible "constraint" we can really specify is
+		the branch name if the branch name is not HEAD. Otherwise, it's a specific revision.
+
+		However, if we can infer a tag that points to the revision or the branch, we may be able
+		to use that as the constraint
+	*/
+
+	pc.Ident = gps.ProjectIdentifier{ProjectRoot: gps.ProjectRoot(pkg.Importpath), Source: pkg.Repository}
+
+	// The default constraint is just a revision
+	var revision = gps.Revision(pkg.Revision)
+
+	// See if we can get a version from that constraint
+	version, err := lookupVersionForLockedProject(pc.Ident, revision, revision, i.sm)
```

Comment:
The second parameter `c gps.Constraint` should be `nil`, instead of passing in the revision again.

The purpose of the constraint parameter is to help pick the appropriate tag when a revision has multiple tags, by filtering with the semver constraint from the manifest (which we don't have with gb). To be clear, it doesn't cause the function to fail, but we know that since it's a revision, it's never going to help in that case. 😀